packages/contracts/contracts/verifierHelpers/BaseVerifierHelper.sol

```diff
@@ -8,64 +8,95 @@ contract BaseVerifierHelper {
+    uint256 public immutable chainid;
 
     struct EpochKeySignals {
-        bool revealNonce;
-        uint8 nonce;
-        uint48 epoch;
-        uint160 attesterId;
-        uint256 stateTreeRoot;
         uint256 epochKey;
+        uint256 stateTreeRoot;
         uint256 data;
+        uint160 attesterId;
+        uint48 epoch;
+        uint48 chainId;
+        uint8 nonce;
+        bool revealNonce;
     }
 
     struct ReputationSignals {
-        bool proveGraffiti;
-        bool proveMinRep;
-        bool proveMaxRep;
-        bool proveZeroRep;
-        bool revealNonce;
-        uint8 nonce;
-        uint48 epoch;
-        uint160 attesterId;
-        uint256 stateTreeRoot;
         uint256 epochKey;
-        uint256 graffiti;
+        uint256 stateTreeRoot;
         uint256 minRep;
         uint256 maxRep;
+        uint256 graffiti;
+        uint256 data;
+        uint160 attesterId;
+        uint48 epoch;
+        uint48 chainId;
+        uint8 nonce;
+        bool revealNonce;
+        bool proveMinRep;
+        bool proveMaxRep;
+        bool proveZeroRep;
+        bool proveGraffiti;
     }
 
     error AttesterInvalid();
     error InvalidEpochKey();
     error InvalidProof();
     error CallerInvalid();
+    error ChainIdNotMatch(uint48 chainId);
 
     function decodeEpochKeyControl(
         uint256 control
     )
         public
         pure
         returns (
-            bool revealNonce,
-            uint160 attesterId,
+            uint8 nonce,
             uint48 epoch,
-            uint8 nonce
+            uint160 attesterId,
+            bool revealNonce,
+            uint48 chainId
         )
     {
-        uint8 attesterIdBits = 160;
-        uint8 epochBits = 48;
         uint8 nonceBits = 8;
-        revealNonce =
-            ((control >> (attesterIdBits + epochBits + nonceBits)) & 1) != 0;
-        attesterId = uint160(
-            (control >> (epochBits + nonceBits)) & ((1 << attesterIdBits) - 1)
+        uint8 epochBits = 48;
+        uint8 attesterIdBits = 160;
+        uint8 revealNonceBit = 1;
+        uint8 chainIdBits = 36;
+        uint8 accBits = 0;
+
+        nonce = uint8(shiftAndParse(control, accBits, nonceBits));
+        accBits += nonceBits;
+
+        epoch = uint48(shiftAndParse(control, accBits, epochBits));
+        accBits += epochBits;
+
+        attesterId = uint160(shiftAndParse(control, accBits, attesterIdBits));
+        accBits += attesterIdBits;
+
+        revealNonce = bool(
+            shiftAndParse(control, accBits, revealNonceBit) != 0
         );
-        epoch = uint48((control >> nonceBits) & ((1 << epochBits) - 1));
-        nonce = uint8(control & ((1 << nonceBits) - 1));
-        return (revealNonce, attesterId, epoch, nonce);
+        accBits += revealNonceBit;
+
+        chainId = uint48(shiftAndParse(control, accBits, chainIdBits));
+        return (nonce, epoch, attesterId, revealNonce, chainId);
```

Comment:
I think if you use named variable in the function definition, like `returns (uint varA, uint varB)`, you don't need to explicitly return the value. So here this line can be removed.

Reply:
Good suggestion 🙏🏻
[d10d36b](https://github.com/Unirep/Unirep/pull/547/commits/d10d36b975a3c551d8fd9a9bdcb2d49eec190379)